lib/tasks/manageiq/cypress.rake

```diff
@@ -0,0 +1,69 @@
+namespace :cypress do
+  desc "Run cypress tests"
+  task :ui => 'ui:run'
+
+  namespace :ui do
+    desc "Seed database/configure assets for cypress specs"
+    task :seed do |rake_task|
+      app_prefix = rake_task.name.chomp('cypress:ui:seed')
+      Rake::Task["#{app_prefix}integration:seed"].invoke
+    end
+
+    desc "Interactively run cypress tests"
+    task :open => ['ui:setup'] do
+      sh "#{yarn_cmd} cypress open"
+    end
+
+    desc "Run headless cypress tests"
+    task :run => ['ui:setup'] do
+      sh "#{yarn_cmd} cypress run --headless --browser firefox --config video=false"
+    end
+
+    desc "Stop the backend Rails server"
+    task :stop do |rake_task|
+      app_prefix = rake_task.name.chomp('cypress:ui:stop')
+      Rake::Task["#{app_prefix}integration:stop_server"].invoke
+    end
+
+    task :setup do |rake_task|
+      app_prefix = rake_task.name.chomp('cypress:ui:setup')
+      Rake::Task["#{app_prefix}integration:start_server"].invoke
+    end
+
+    desc "Run cypress tests in 'development' mode"
+    task :dev => 'dev:run'
+
+    # Aliases for running with CYPRESS_DEV
+    namespace :dev do
+      desc "'Development Mode' for cypress:ui:seed"
+      task :seed => [:env, 'cypress:ui:seed']
+
+      desc "'Development Mode' for cypress:ui:open"
+      task :open => [:env, 'cypress:ui:open']
+
+      desc "'Development Mode' for cypress:ui:run"
+      task :run => [:env, 'cypress:ui:run']
+
+      desc "'Development Mode' for cypress:ui:setup"
+      task :setup => [:env, 'cypress:ui:setup']
+
+      desc "'Development Mode' for cypress:ui:stop"
+      task :stop => [:env, 'cypress:ui:stop']
+
+      # Helper task to defaulte the CYPRESS_DEV env var to "1"
+      task :env do
+        ENV["CYPRESS_DEV"] = "1"
+      end
+    end
+
+    def yarn_cmd
+      cmd = "yarn"
+
+      unless defined?(ENGINE_ROOT)
+        cmd += " --cwd #{File.expand_path(File.join("..", "..", ".."), __dir__)}"
```

Comment:
this only works when you have `manageiq/plugins/manageiq-ui-classic`, right?  Better to use `spec/manageiq`?

Reply:
> (That said, maybe I just don't know when you wouldn't have ENGINE_ROOT.)

This allows it to be run from the `ManageIQ/manageiq` repo instead of from this repo.  Other wise the `node_modules` dir can't be found when running from that project.

However, we could probably just get away always doing this and just making it a constant, since it doesn't really hurt anything having the extra `--cwd` flag there.  So something like this at the top of the file:

```ruby
CYPRESS_YARN_CMD = "yarn --cwd #{File.expand_path(File.join("..", "..", ".."), __dir__)}"
```

Reply:
Oh, I remember why I didn't use a constant now...

Because this is in `lib/tasks/**/*.rake` _**AND**_ this exists in our `Rakefile`:

https://github.com/ManageIQ/manageiq-ui-classic/blob/b6496c8ada278b1387d424e37aa3a6bb21903452/Rakefile#L6

It means that we are loading this task twice in two contexts:

- once in the top level, creating the `cypress:ui:*` tasks
- once in the `app:*` context, creating the `app:cypress:ui:*` tasks

This is probably "wrong" from a Rails/`rake` standards standpoint, but since I don't know if there is a reason for it's existence, and the following warnings urk me:

```
./manageiq-ui-classic/lib/tasks/manageiq/cypress.rake:1: warning: already initialized constant CYPRESS_YARN_CMD
./manageiq-ui-classic/lib/tasks/manageiq/cypress.rake:1: warning: previous definition of CYPRESS_YARN_CMD was here
```

I am probably just going to leave it as is then.  This file might turned into a `RakeLib` library that other repos can reuse, so it will probably will be cleaned up in the future.

Reply:
I don't think you've answered my question here, what's the situation this is needed for? 

Are you saying that if I `cd manageiq/ ; rake cypress:open`, yarn will fail because it tries to run it in ui-classic, and adding `../../..` forces yarn to use `manageiq/node_modules`?

If so, I think it's the opposite of what we should be doing.

`Dir.chdir ManageIQ::UI::Classic::Engine.root` is what we're using for the other tasks that need to be run in the UI.

(And for the record, I've never ever had a need to use an `app:` prefixed rake task.)

Reply:
> (And for the record, I've never ever had a need to use an app: prefixed rake task.)

For the record, anything in `lib/tasks/*.rake` is exposed to the main Rails application the engine is included in (in this case, `ManageIQ/manageiq`), and the `app:` prefix is how from the Engine repo you can execute it normally is if it was from a repo.

As mentioned above, the reason why you might _**not**_ need to use the `app:` prefix is because of this line:

https://github.com/ManageIQ/manageiq-ui-classic/blob/b6496c8ada278b1387d424e37aa3a6bb21903452/Rakefile#L6

Which also includes the rake tasks in their original namespace.

You can see that both of these tasks exist for yourself by running the below from this repo:

```console
$ rake -AT update:ui
```

(note:  `-T` lists all the tasks, and providing an argument, `update:ui`, allows for an internal "grep", and -A will also include tasks that don't have descriptions, `desc`, which 
`update:ui` does not)


> Are you saying that if I cd manageiq/ ; rake cypress:open, yarn will fail because it tries to run it in ui-classic, and adding ../../.. forces yarn to use manageiq/node_modules?
> 
> If so, I think it's the opposite of what we should be doing.

The opposite.  If you don't include the `--cwd` flag and you try running the tasks from `manageiq`, it will fail because it tries running `yarn` from `manageiq` directory, and not `manageiq-ui-classic`, and there is "`yarn` files" there for it to pick up.

There was an argument that maybe it makes sense to instead only support these tasks from the `Engine` only, and not allow them to be run from both `manageiq` and `manageiq-ui-classic`.  I decided to support both for now, since say you were working on a `manageiq` model feature, and you just wanted to confirm things were still working with the UI specs, this would save you having to run things from the `manageiq-ui-classic` repo.  But again, that is arguably not worth the time as well.